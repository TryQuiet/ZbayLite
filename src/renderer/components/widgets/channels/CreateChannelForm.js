import React from 'react'
import PropTypes from 'prop-types'
import * as R from 'ramda'
import { Formik, Form } from 'formik'
import * as Yup from 'yup'

import Grid from '@material-ui/core/Grid'
import { withStyles } from '@material-ui/core/styles'
import WarningIcon from '@material-ui/icons/Warning'
import { Typography } from '@material-ui/core'

import TextField from '../../ui/form/TextField'
import LoadingButton from '../../ui/LoadingButton'

const styles = theme => ({
  fullContainer: {
    width: '100%',
    height: '100%'
  },
  gutter: {
    marginTop: 8,
    marginBottom: 24
  },
  button: {
    width: 165,
    backgroundColor: theme.palette.colors.zbayBlue,
    color: theme.palette.colors.white,
    '&:hover': {
      backgroundColor: theme.palette.colors.zbayBlue
    },
    textTransform: 'none',
    height: 48,
    fontWeight: 'normal'
  },
  title: {
    marginBottom: 24
  },
  warrningIcon: {
    color: '#FFCC00'
  },
  iconDiv: {
    width: 24,
    height: 28,
    marginRight: 8
  },
  warrningMessage: {
    wordBreak: 'break-word'
  }
})
export const parseChannelName = (name = '') => {
  return name.toLowerCase().replace(/ +/g, '-')
}
export const formSchema = Yup.object().shape({
  name: Yup.string()
    .max(20, 'Channel name is too long.')
    .test(
      'testFormat',
      'Channel name can contain only small characters and up to one hyphen.',
      function (value) {
        return parseChannelName(value).match(/^[a-z0-9]+([\s-][a-z0-9]+){0,}$/)
      }
    )
    .required('Your channel must have a name.')
})

export const showParsedMessage = (message = '') => {
  return message.includes(' ') || message !== message.toLowerCase()
}

export const CreateChannelForm = ({ classes, onSubmit, setStep }) => (
  <Formik
    validationSchema={formSchema}
    onSubmit={(values, formActions) => {
      onSubmit(
        { ...values, name: parseChannelName(values.name) },
        formActions,
        setStep
      )
    }}
    initialValues={{ name: '' }}
  >
    {({ isSubmitting, values, isValid }) => (
      <Form className={classes.fullContainer}>
        <Grid
          container
          justify='flex-start'
          direction='column'
          className={classes.fullContainer}
        >
          <Typography variant='h3' className={classes.title}>
            Create a new private channel
          </Typography>
          <Typography variant='body2'>Channel name</Typography>
          <TextField name='name' placeholder='my-channel' />
          <div className={classes.gutter}>
            {showParsedMessage(values.name) && isValid && (
              <Grid container alignItems='center' direction='row'>
                <Grid item className={classes.iconDiv}>
                  <WarningIcon className={classes.warrningIcon} />
                </Grid>
                <Grid item xs className=''>
                  <Typography
                    variant='body2'
                    className={classes.warrningMessage}
                  >
                    Your channel will be created as{' '}
                    {parseChannelName(values.name)}
                  </Typography>
                </Grid>
              </Grid>
            )}
          </div>
          <LoadingButton
            className={classes.button}
            variant='contained'
            color='primary'
            disabled={isSubmitting || !isValid}
            inProgress={isSubmitting}
            type='submit'
            text='Create Channel'
          />
        </Grid>
      </Form>
    )}
  </Formik>
)

CreateChannelForm.propTypes = {
  classes: PropTypes.object.isRequired,
  onSubmit: PropTypes.func.isRequired,
  setStep: PropTypes.func.isRequired
}

export default R.compose(React.memo, withStyles(styles))(CreateChannelForm)
