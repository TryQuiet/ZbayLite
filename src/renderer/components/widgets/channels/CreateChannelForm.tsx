import React from 'react'

import Grid from '@material-ui/core/Grid'
import { makeStyles } from '@material-ui/core/styles'
import { Typography } from '@material-ui/core'

import LoadingButton from '../../ui/LoadingButton/LoadingButton'
import { TextInput } from '../../../forms/components/textInput'
import { useForm } from 'react-hook-form'
import { channelNameField } from '../../../forms/fields/createUserFields'

const useStyles = makeStyles((theme) => ({
  fullContainer: {
    width: '100%',
    height: '100%'
  },
  gutter: {
    marginTop: 8,
    marginBottom: 24
  },
  button: {
    width: 165,
    backgroundColor: theme.palette.colors.zbayBlue,
    color: theme.palette.colors.white,
    '&:hover': {
      backgroundColor: theme.palette.colors.zbayBlue
    },
    textTransform: 'none',
    height: 48,
    fontWeight: 'normal'
  },
  title: {
    marginBottom: 24
  },
  warrningIcon: {
    color: '#FFCC00'
  },
  iconDiv: {
    width: 24,
    height: 28,
    marginRight: 8
  },
  warrningMessage: {
    wordBreak: 'break-word'
  }
}))

export const showParsedMessage = (message = '') => {
  return message.includes(' ') || message !== message.toLowerCase()
}

interface CreateChannelFormProps {
  // onSubmit: ({ name }: { name: string }, formActions, setStep) => void
  setStep: (arg: number) => void
}

interface CreateChannelValues {
  channelName: string
}

const channelFields = {
  channelName: channelNameField()
}

export const CreateChannelForm: React.FC<CreateChannelFormProps> = ({ setStep }) => {
  const classes = useStyles({})

  const { register, handleSubmit, formState: { errors } } = useForm<CreateChannelValues>({
    mode: 'onTouched'
  })

  const onSubmit = () => {
    // (values, formActions) => {
    //   onSubmit(
    //     { ...values, name: parseChannelName(values.name) },
    //     formActions,
    //     setStep
    //   )
    // }
    setStep(1)
  }

  return (
    <form
      onSubmit={handleSubmit(onSubmit)}
    >
      <Grid
        container
        justify='flex-start'
        direction='column'
        className={classes.fullContainer}
      >
        <Typography variant='h3' className={classes.title}>
          Create a new private channel (temporarily disabled)
        </Typography>
        <Typography variant='body2'>Channel name</Typography>
        <TextInput
          name='name'
          defaultValue={''}
          variant='outlined'
          errors={errors}
          classes={''}

          {...channelFields.channelName.fieldProps}
          fullWidth
          {...register('channelName', channelFields.channelName.validation)}
          inputRef={register('channelName', channelFields.channelName.validation).ref}

          placeholder={'Enter a channel name'}
        />
        <div className={classes.gutter}>
          {/* {true && ( // disabled for now
            <Grid container alignItems='center' direction='row'>
              <Grid item className={classes.iconDiv}>
                <WarningIcon className={classes.warrningIcon} />
              </Grid>
              <Grid item xs className=''>
                <Typography
                  variant='body2'
                  className={classes.warrningMessage}
                >
                  Your channel will be created as{` ${}`}

                </Typography>
              </Grid>
            </Grid>
          )} */}
        </div>
        <LoadingButton
          variant='contained'
          color='primary'
          disabled={false} // Temporarily turn off
          inProgress={false}
          type='submit'
          text='Create Channel'
          classes={{ button: classes.button }}
        />
      </Grid>
    </form>
  )
}

export default CreateChannelForm
