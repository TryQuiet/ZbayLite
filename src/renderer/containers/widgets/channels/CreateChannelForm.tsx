import React from 'react'
import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'

import CreateChannelFormComponent from '../../../components/widgets/channels/CreateChannelForm'
import CreateChannelFromFinish from '../../../components/widgets/channels/CreateChannelFormFinish'
// import channelsHandlers from '../../../store/handlers/channels'

export const mapDispatchToProps = dispatch =>
  bindActionCreators(
    {
      // onSubmit: channelsHandlers.epics.createChannel
    },
    dispatch
  )
const stepToComponent = {
  0: CreateChannelFormComponent,
  1: CreateChannelFromFinish
}
export const CreateChannelForm = ({ ...props }) => {
  const [step, setStep] = React.useState(0)
  const Component = stepToComponent[step]
  return <Component {...props} setStep={setStep} />
}

export default connect(null, mapDispatchToProps)(CreateChannelForm)
