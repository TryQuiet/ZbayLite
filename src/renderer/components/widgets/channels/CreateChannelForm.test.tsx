import React from 'react'
import { shallow } from 'enzyme'

import { CreateChannelForm } from './CreateChannelForm'

import { MuiThemeProvider } from '@material-ui/core'
import theme from '../../../theme'

describe('CreateChannelForm', () => {
  it('renders component', () => {
    const result = shallow(
      <MuiThemeProvider theme={theme}>
        <CreateChannelForm
          onSubmit={jest.fn()}
          setStep={jest.fn()}
        />
      </MuiThemeProvider>
    )
    expect(result).toMatchSnapshot()
  })
})
