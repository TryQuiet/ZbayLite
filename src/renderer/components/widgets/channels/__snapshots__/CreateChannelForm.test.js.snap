// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateChannelForm renders component 1`] = `
<Formik
  enableReinitialize={false}
  initialValues={
    Object {
      "name": "",
    }
  }
  isInitialValid={false}
  onSubmit={[Function]}
  validateOnBlur={true}
  validateOnChange={true}
  validationSchema={
    ObjectSchema {
      "_blacklist": RefSet {
        "list": Set {},
        "refs": Map {},
      },
      "_conditions": Array [],
      "_defaultDefault": [Function],
      "_deps": Array [],
      "_excludedEdges": Array [],
      "_exclusive": Object {},
      "_mutate": undefined,
      "_nodes": Array [
        "name",
      ],
      "_options": Object {
        "abortEarly": true,
        "recursive": true,
      },
      "_type": "object",
      "_typeError": [Function],
      "_whitelist": RefSet {
        "list": Set {},
        "refs": Map {},
      },
      "fields": Object {
        "name": StringSchema {
          "_blacklist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {
            "required": true,
            "testFormat": false,
          },
          "_mutate": undefined,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "string",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "tests": Array [
            [Function],
            [Function],
          ],
          "transforms": Array [
            [Function],
          ],
        },
      },
      "tests": Array [],
      "transforms": Array [
        [Function],
      ],
    }
  }
>
  <Component />
</Formik>
`;
